src/cargo/sources/registry/mod.rs

```diff
@@ -709,26 +709,33 @@ impl<'cfg> Source for RegistrySource<'cfg> {
         kind: QueryKind,
         f: &mut dyn FnMut(Summary),
     ) -> Poll<CargoResult<()>> {
-        // If this is a precise dependency, then it came from a lock file and in
+        let mut req = dep.version_req().clone();
+
+        // Handle `cargo update --precise` here.
+        if let Some((_, requested)) = self
+            .source_id
+            .precise_registry_version(dep.package_name().as_str())
+            .filter(|(c, _)| req.matches(c))
+        {
+            req.lock_to(&requested);
+        }
+
+        // If this is a locked dependency, then it came from a lock file and in
         // theory the registry is known to contain this version. If, however, we
         // come back with no summaries, then our registry may need to be
         // updated, so we fall back to performing a lazy update.
-        if kind == QueryKind::Exact && dep.source_id().precise().is_some() && !self.ops.is_updated()
-        {
+        if kind == QueryKind::Exact && req.is_locked() && !self.ops.is_updated() {
```

Comment:
I think `req.is_locked()` and `dep.source_id().precise().is_some()` have the same meaning. The same method sets both of them, and assert_eq pass all tests. So what is the change for clarity.

Reply:
I feel I'm missing something with what this comment is referring to.  Where is `dep.source_id().precise().is_some()` that this is referring to?

Reply:
The code before the PR.

Reply:
Ah, hidden scroll bars...